[6.1] Add file types for media custom field plugin.

sergeytolkachyov opened the pull request:
When creating a custom field, do you want to be able to select not only images, but also documents, videos, and audio? This PR adds the ability to specify one or more file types for a custom media type field.

### Summary of Changes
- The `file types` parameter was added to the field parameters during creation.
- There are 4 file types: `images`, `audios`, `videos`, `documents`. This list decides which of the allowed file extensions from Media Manager configuration are used. 
- `images` file type is selected by default and for empty parameter value
- You can select one or several file types: only `videos` and `documents` for example
- new attribute for Accessiblemedia Form Field: `types`. Similar to Media field.

### Testing Instructions
- Apply this PR
- Create a 4 new custom fields for articles or contacts.
<img width="812" height="560" alt="3 new custom field types" src="https://github.com/user-attachments/assets/e3b1a346-1a8f-407e-9c43-5818669ebb74" />


- Look at new field types:
-- `Audio (media_audio)`
-- `Document (media_document)`
-- `Video (media_video)`
<img width="861" height="904" alt="image" src="https://github.com/user-attachments/assets/9ba46a9a-765b-47d8-ad5e-90e37781cde9" />
<img width="887" height="979" alt="image" src="https://github.com/user-attachments/assets/de47f7aa-73b0-46c8-8064-f04d5819d46f" />

- save your fields: `Image` with type `media`, `Audio` with type `Audio (media_audio)`, `Document` with type `Document (media_document)` and `Video` with type `Video (media_video)`
- go to article edit view and try to choose a media file in all of these fields - You'll be able to select only corresponding file types.
- See also a modal window title is `Change image` for `Media (images)` and `Change file` for other file types.
![image](https://github.com/user-attachments/assets/306f862b-48df-49fc-a57c-cbebd135435c)
<img width="758" height="497" alt="image" src="https://github.com/user-attachments/assets/e973f5eb-e773-4377-9a12-90edf8047977" />

- Check that download link for selected file has been rendered successfully in the frontend for `documents` file types. The link text is `Download {filename}` by default. Shortcode `{filename}` is replaced by current file name. You can specify your own text here.
Backend
<img width="1221" height="381" alt="image" src="https://github.com/user-attachments/assets/fbd80851-84a2-4df8-b517-d3b4c143701e" />
Frontend
<img width="696" height="203" alt="image" src="https://github.com/user-attachments/assets/cf61b998-ef4d-44e6-8b5c-1034be05ead3" />

- Check that `<video>` tag for selected file has been rendered successfully in the frontend for `video` file type. 
<img width="829" height="405" alt="image" src="https://github.com/user-attachments/assets/501c5d91-bef3-4528-a320-6d932f677302" />

- If you're enabled a `poster` for video - check the additional field for poster image in admin panel and `<video>` tag for selected file has been rendered successfully in the frontend with specified poster image.
Backend
<img width="1182" height="520" alt="image" src="https://github.com/user-attachments/assets/ae506cef-4fa3-4ea0-b1c7-3667f01c8507" />
Frontend
<img width="998" height="978" alt="image" src="https://github.com/user-attachments/assets/963bf1dd-1719-4086-a468-ccda3afdfd6a" />

- Check that `<audio>` tag for selected file has been rendered successfully in the frontend for `audio` file type. 
Backend
<img width="1208" height="401" alt="image" src="https://github.com/user-attachments/assets/27918bca-23f9-4640-81c4-99e3f24442a3" />
Frontend
<img width="986" height="440" alt="image" src="https://github.com/user-attachments/assets/8e66f1cf-6a61-4eb6-9da9-8d34ac86c321" />

- Select `document` file and check that link to download selected file has been rendered successfully in the frontend. Make sure that the link text matches the one specified in the field `link text` and shortcode `{filename}` has been successfully replaced with file name.
Backend
<img width="1249" height="499" alt="image" src="https://github.com/user-attachments/assets/def13a5d-8ef2-4cf5-b2ab-4f5a83e76c96" />
Frontend
<img width="557" height="329" alt="image" src="https://github.com/user-attachments/assets/582d3840-f957-4cfa-ad52-7dd7e37f7309" />

### Actual result BEFORE applying this Pull Request
You cannot select anything except images with `media` custom field.

### Expected result AFTER applying this Pull Request
Now you can select 4 file types for custom field type media. You can select a `mp4` or `pdf` in your media custom field. 
- For `image` files use a standard `Media (media)` field type. If image file has been selected - you can use `alt text` and `empty alt` fields. Image will render in frontend.
- For `audio` or `video` file types use new `Audio (media_audio)` or `Video (media_video)` field types - `<audio>` or `<video>` tag will render in frontend. You can specifiy a poster image for video if you need. You can specifiy HTML5 attributes like `controls`, `muted`, `loop` etc.
- For `document` file type use new `Document (media_document)` field type.   Download link will render in frontend. You can specify `link text` for download link.

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [ ] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

sergeytolkachyov committed:
Add file types for media custom field plugin.

brianteeman commented on administrator/language/en-GB/plg_fields_media.ini:
```diff
@@ -10,3 +10,9 @@ PLG_FIELDS_MEDIA_PARAMS_IMAGE_CLASS_LABEL="Image Class"
 PLG_FIELDS_MEDIA_PARAMS_PREVIEW_INLINE="Inline"
 PLG_FIELDS_MEDIA_PARAMS_PREVIEW_LABEL="Preview"
 PLG_FIELDS_MEDIA_XML_DESCRIPTION="This plugin lets you create new fields of type 'media' in any extensions where custom fields are supported."
+PLG_FIELDS_MEDIA_PARAMS_TYPES_LABEL = "File types"
+PLG_FIELDS_MEDIA_PARAMS_TYPES_DESCRIPTION = "You can configure a file types in Media manager settings"
+PLG_FIELDS_MEDIA_PARAMS_TYPES_IMAGES = "Images"
+PLG_FIELDS_MEDIA_PARAMS_TYPES_AUDIOS = "Audios"
+PLG_FIELDS_MEDIA_PARAMS_TYPES_VIDEOS = "Videos"
+PLG_FIELDS_MEDIA_PARAMS_TYPES_DOCUMENTS = "Documents"
```
instead of adding the language strings at the end they should be inserted into the entire list of strings so that they are all alphabetically sorted

brianteeman replied:
To follow our style guides for labels the label should be changed to
`PLG_FIELDS_MEDIA_PARAMS_TYPES_LABEL = "File Types"`

brianteeman replied:
Audios and Videos just seems wrong english. Perhaps it should be "Audio Files" and "Video Files" or just "Audio" and "Video". Maybe another native english speaker has an opinion

sergeytolkachyov replied:
> To follow our style guides for labels the label should be changed to `PLG_FIELDS_MEDIA_PARAMS_TYPES_LABEL = "File Types"`

Done

sergeytolkachyov replied:
> Audios and Videos just seems wrong english. Perhaps it should be "Audio Files" and "Video Files" or just "Audio" and "Video". Maybe another native english speaker has an opinion

Changed to `audio` and `video`.

brianteeman commented on plugins/fields/media/params/media.xml:
```diff
@@ -27,6 +27,20 @@
 				validate="CssIdentifier"
 			/>
+
+			<field type="list"
+				   name="types"
+				   label="PLG_FIELDS_MEDIA_PARAMS_TYPES_LABEL"
+				   description="PLG_FIELDS_MEDIA_PARAMS_TYPES_DESCRIPTION"
+				   default="images"
+				   multiple="true"
+				   layout="joomla.form.field.list-fancy-select"
+			>
+				<option value="images">PLG_FIELDS_MEDIA_PARAMS_TYPES_IMAGES</option>
+				<option value="audios">PLG_FIELDS_MEDIA_PARAMS_TYPES_AUDIOS</option>
+				<option value="videos">PLG_FIELDS_MEDIA_PARAMS_TYPES_VIDEOS</option>
+				<option value="documents">PLG_FIELDS_MEDIA_PARAMS_TYPES_DOCUMENTS</option>
+			</field>
```
```suggestion
			<field type="list"
				name="types"
				label="PLG_FIELDS_MEDIA_PARAMS_TYPES_LABEL"
				description="PLG_FIELDS_MEDIA_PARAMS_TYPES_DESCRIPTION"
				default="images"
				multiple="true"
				layout="joomla.form.field.list-fancy-select"
				>
				<option value="images">PLG_FIELDS_MEDIA_PARAMS_TYPES_IMAGES</option>
				<option value="audios">PLG_FIELDS_MEDIA_PARAMS_TYPES_AUDIOS</option>
				<option value="videos">PLG_FIELDS_MEDIA_PARAMS_TYPES_VIDEOS</option>
				<option value="documents">PLG_FIELDS_MEDIA_PARAMS_TYPES_DOCUMENTS</option>
			</field>
```

sergeytolkachyov committed:
Update plugins/fields/media/params/media.xml

Co-authored-by: Brian Teeman <brian@teeman.net>

sergeytolkachyov committed:
Fix language string and strings sorting

sergeytolkachyov committed:
Merge remote-tracking branch 'origin/5.3-dev' into 5.3-dev

QuyTon commented on libraries/src/Form/Field/AccessiblemediaField.php:
```diff
@@ -68,6 +68,14 @@ class AccessiblemediaField extends SubformField
+    /**
+     * Media types like 'images, audios, documents, videos'
+     *
+     * @var string
+     * @since 5.3.0
```
```suggestion
     * @var    string
     * @since  __DEPLOY_VERSION__
```

sergeytolkachyov committed:
Update libraries/src/Form/Field/AccessiblemediaField.php

Co-authored-by: Quy Ton <quy@nomonkeybiz.com>

Septdir commented:
I have tested this item :white_check_mark: successfully on 9a36d9b3029c73d4be420efe80b87fcc6b20511b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/45013">issues.joomla.org/tracker/joomla-cms/45013</a>.</sub>

brianteeman commented on administrator/language/en-GB/plg_fields_media.ini:
```diff
@@ -9,4 +9,10 @@ PLG_FIELDS_MEDIA_PARAMS_DIRECTORY_LABEL="Directory"
+PLG_FIELDS_MEDIA_PARAMS_TYPES_AUDIOS = "Audio"
+PLG_FIELDS_MEDIA_PARAMS_TYPES_DOCUMENTS = "Documents"
+PLG_FIELDS_MEDIA_PARAMS_TYPES_IMAGES = "Images"
+PLG_FIELDS_MEDIA_PARAMS_TYPES_VIDEOS = "Video"
```
```suggestion
PLG_FIELDS_MEDIA_PARAMS_TYPES_AUDIOS = "Audio"
PLG_FIELDS_MEDIA_PARAMS_TYPES_DESCRIPTION = "You can configure a file types in Media manager settings"
PLG_FIELDS_MEDIA_PARAMS_TYPES_DOCUMENTS = "Documents"
PLG_FIELDS_MEDIA_PARAMS_TYPES_IMAGES = "Images"
PLG_FIELDS_MEDIA_PARAMS_TYPES_LABEL = "File Types"
PLG_FIELDS_MEDIA_PARAMS_TYPES_VIDEOS = "Video"
```

sergeytolkachyov replied:
What's the reason for this sorting?

brianteeman replied:
the strings should be listed in alphabetical order

hans2103 commented:
I have tested this item :white_check_mark: successfully on 9a36d9b3029c73d4be420efe80b87fcc6b20511b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/45013">issues.joomla.org/tracker/joomla-cms/45013</a>.</sub>

sergeytolkachyov committed:
Update administrator/language/en-GB/plg_fields_media.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

hans2103 commented:
I have tested this item :white_check_mark: successfully on dff217abe7a8ff3171b2cf75a4c357b202b27e3f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/45013">issues.joomla.org/tracker/joomla-cms/45013</a>.</sub>

sergeytolkachyov commented:
@hans2103 can you test again, please?

gug2 commented:
I have tested this item :white_check_mark: successfully on dff217abe7a8ff3171b2cf75a4c357b202b27e3f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/45013">issues.joomla.org/tracker/joomla-cms/45013</a>.</sub>

viocassel commented:
I have tested this item :white_check_mark: successfully on dff217abe7a8ff3171b2cf75a4c357b202b27e3f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/45013">issues.joomla.org/tracker/joomla-cms/45013</a>.</sub>

QuyTon commented on plugins/fields/media/params/media.xml:
```diff
@@ -27,6 +27,20 @@
 				validate="CssIdentifier"
 			/>
+
+			<field type="list"
+				name="types"
+				label="PLG_FIELDS_MEDIA_PARAMS_TYPES_LABEL"
```
```suggestion
			<field
				name="types"
				type="list"
				label="PLG_FIELDS_MEDIA_PARAMS_TYPES_LABEL"
```

QuyTon commented on administrator/language/en-GB/plg_fields_media.ini:
```diff
@@ -9,4 +9,10 @@ PLG_FIELDS_MEDIA_PARAMS_DIRECTORY_LABEL="Directory"
 PLG_FIELDS_MEDIA_PARAMS_PREVIEW_INLINE="Inline"
 PLG_FIELDS_MEDIA_PARAMS_PREVIEW_LABEL="Preview"
+PLG_FIELDS_MEDIA_PARAMS_TYPES_AUDIOS = "Audio"
+PLG_FIELDS_MEDIA_PARAMS_TYPES_DESCRIPTION = "You can configure a file types in Media manager settings"
```
I don't understand which settings can be configured. Please explain. Thanks.

sergeytolkachyov replied:
> I don't understand which settings can be configured. Please explain. Thanks.

Files extensions and mime types in Media manager settings. 
![image](https://github.com/user-attachments/assets/49092275-9046-4bcb-ac4b-54f6076d27c1)

sergeytolkachyov committed:
Update plugins/fields/media/params/media.xml

Co-authored-by: Quy Ton <quy@nomonkeybiz.com>

sergeytolkachyov commented:
I have tested this item :white_check_mark: successfully on 8a2e00b270c49afc4a5a7fc8404228ba5d2c77ea<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/45013">issues.joomla.org/tracker/joomla-cms/45013</a>.</sub>

web-eau-net commented:
I have tested this item :white_check_mark: successfully on 8a2e00b270c49afc4a5a7fc8404228ba5d2c77ea<br /><br />Nice job!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/45013">issues.joomla.org/tracker/joomla-cms/45013</a>.</sub>

QuyTon commented on administrator/language/en-GB/plg_fields_media.ini:
```diff
@@ -9,4 +9,10 @@ PLG_FIELDS_MEDIA_PARAMS_DIRECTORY_LABEL="Directory"
 PLG_FIELDS_MEDIA_PARAMS_PREVIEW_INLINE="Inline"
 PLG_FIELDS_MEDIA_PARAMS_PREVIEW_LABEL="Preview"
+PLG_FIELDS_MEDIA_PARAMS_TYPES_AUDIOS = "Audio"
+PLG_FIELDS_MEDIA_PARAMS_TYPES_DESCRIPTION = "You can configure a file types in Media manager settings"
```
```suggestion
PLG_FIELDS_MEDIA_PARAMS_TYPES_DESCRIPTION = "Configure file types in Media manager settings."
```
```suggestion
PLG_FIELDS_MEDIA_PARAMS_TYPES_DESCRIPTION = "Configure file types in Media settings."
```

brianteeman commented:
I am sorry but unless I misunderstand something this creates a very **confusing** UI in the admin and **broken** output in the front end and I dont see how any of you can have tested this successfully

![image](https://github.com/user-attachments/assets/b47e0e70-0be4-4919-b771-8a978fdd60f9)

![image](https://github.com/user-attachments/assets/f16506a5-e3f3-4d46-9cb2-cd37a73ba912)

![image](https://github.com/user-attachments/assets/f3e2f1ed-73c6-4453-936e-2263a980d805)

![image](https://github.com/user-attachments/assets/35ab3654-5b5e-4a7d-a220-f1c59e46939f)

brianteeman commented:
I have tested this item :red_circle: unsuccessfully on 8a2e00b270c49afc4a5a7fc8404228ba5d2c77ea<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/45013">issues.joomla.org/tracker/joomla-cms/45013</a>.</sub>

brianteeman commented:
@sergeytolkachyov I have removed the successful test result you posted as it is assumed that the author would not submit something that didnt work and only tests from other people are valid

brianteeman commented:
When testing this PR you need to actually test what the field produces and not the limited tests in the original post. In summary you might be able to select a document or a video but this field will NOT be able to display the document or video<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/45013">issues.joomla.org/tracker/joomla-cms/45013</a>.</sub>

sergeytolkachyov commented:
> When testing this PR you need to actually test what the field produces and not the limited tests in the original post. In summary you might be able to select a document or a video but this field will NOT be able to display the document or video
> 
> This comment was created with the [J!Tracker Application](https://github.com/joomla/jissues) at [issues.joomla.org/tracker/joomla-cms/45013](https://issues.joomla.org/tracker/joomla-cms/45013).

I think all the successful tests were related to checking the very possibility of choosing a file type other than the image. in most cases, the standard layouts are redefined and in the frontend, the field is only required to provide the actual path to the file. and the desired layout is already determined by the developer in a specific redefinition. nevertheless, I agree that the output of pdf or mp3 in the image tag will be incorrect. after all, audio and video can be output as a player, and pdf as an <object>, but you can't display docx like that anymore. Such a variety of formats will complicate the logic of choosing a layout. would it be enough to output a file (not an image) as a download link?

brianteeman commented:
For files you are correct it will just have to be a link

sergeytolkachyov committed:
Merge branch 'joomla:5.3-dev' into 5.3-dev

sergeytolkachyov commented:
@brianteeman by default the field `name` attribute is `imagefile`, but if we are choosing a documents it will not an `imagefile` by logic. But I think I will not change a field name because it will be a bc breakes.

sergeytolkachyov committed:
- Rebuild XML constructing with SimpleXMLElement
- add new field - filename
- add language strings for new filename field

sergeytolkachyov committed:
- add `Change file` language string for modal select window
- add changing modal window title depending on the file types

sergeytolkachyov committed:
- change new field name to `linktext`
- add `default` and `hint` attributes for it with language string
- edit media field plugin layout for render links for files and images for images

sergeytolkachyov committed:
language strings sorted alphabetically

sergeytolkachyov committed:
fix namespace place

sergeytolkachyov committed:
fix namespace place

fgsw commented on administrator/language/en-GB/lib_joomla.ini:
```diff
@@ -258,6 +258,7 @@ JLIB_FORM_BUTTON_CLEAR="Clear"
 JLIB_FORM_BUTTON_SELECT="Select"
 JLIB_FORM_CHANGE_IMAGE="Change Image"
 JLIB_FORM_CHANGE_IMAGE_BUTTON="Change Image Button"
+JLIB_FORM_CHANGE_FILE="Change file"
```
Please move up below JLIB_FORM_BUTTON_SELECT="Select"

fgsw commented on language/en-GB/lib_joomla.ini:
```diff
@@ -257,6 +257,7 @@ JLIB_FORM_BUTTON_CLEAR="Clear"
 JLIB_FORM_BUTTON_SELECT="Select"
 JLIB_FORM_CHANGE_IMAGE="Change Image"
 JLIB_FORM_CHANGE_IMAGE_BUTTON="Change Image Button"
+JLIB_FORM_CHANGE_FILE="Change file"
```
Please move up below JLIB_FORM_BUTTON_SELECT="Select"

sergeytolkachyov commented:
@gug2 @hans2103 @viocassel @web-eau-net @brianteeman check it please. Testing instructions has been updated.

sergeytolkachyov committed:
fix code style for continuous-integration/drone/pr

sergeytolkachyov committed:
fix code style for continuous-integration/drone/pr - global_namespace_import

sergeytolkachyov committed:
fix code style for continuous-integration/drone/pr - layouts/joomla/form/field/media.php

sergeytolkachyov committed:
fix code style for continuous-integration/drone/pr - layouts/joomla/form/field/media.php

gug2 commented:
I have tested this item :white_check_mark: successfully on 02796b9fc394e481a1520f39c4aedafdca93af14<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/45013">issues.joomla.org/tracker/joomla-cms/45013</a>.</sub>

brianteeman commented:
still testing but looks good

viocassel commented:
I have tested this item :white_check_mark: successfully on 02796b9fc394e481a1520f39c4aedafdca93af14<br /><br />👍<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/45013">issues.joomla.org/tracker/joomla-cms/45013</a>.</sub>

QuyTon commented on administrator/language/en-GB/lib_joomla.ini:
```diff
@@ -277,6 +278,9 @@ JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_ALT_EMPTY_DESC="Decorative Image -
 JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_ALT_EMPTY_LABEL="No Description"
 JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_ALT_TEXT_LABEL="Image Description (Alt Text)"
 JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_IMAGEFILE_LABEL="Image"
+JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_LINKTEXT = "Link text"
```
Please remove the space before/after `=`.

QuyTon commented on layouts/joomla/form/field/media.php:
```diff
@@ -23,31 +23,31 @@
 /**
  * Layout variables
  * -----------------
- * @var  string   $asset           The asset text
```
Please revert space changes.

sergeytolkachyov replied:
done

sergeytolkachyov committed:
remove the space before/after `=`

QuyTon commented on layouts/joomla/form/field/media.php:
```diff
@@ -148,13 +148,19 @@
     ]);
 }
 
+$imagesOnly = false;
+
+if (count($mediaTypeNames) == 1 && in_array('images', $mediaTypeNames)) {
+    $imagesOnly = true;
+}
```
```suggestion
$imagesOnly = count($mediaTypeNames) == 1 && in_array('images', $mediaTypeNames);
```

sergeytolkachyov committed:
Update layouts/joomla/form/field/media.php

Co-authored-by: Quy Ton <quy@nomonkeybiz.com>

QuyTon commented on administrator/language/en-GB/lib_joomla.ini:
```diff
@@ -256,6 +256,7 @@ JLIB_FILTER_PARAMS_TEXT="Text"
 
 JLIB_FORM_BUTTON_CLEAR="Clear"
 JLIB_FORM_BUTTON_SELECT="Select"
+JLIB_FORM_CHANGE_FILE="Change file"
```
```suggestion
JLIB_FORM_CHANGE_FILE="Change File"
```

sergeytolkachyov replied:
I fixed it now too

QuyTon commented on language/en-GB/lib_joomla.ini:
```diff
@@ -255,6 +255,7 @@ JLIB_FILTER_PARAMS_TEXT="Text"
 
 JLIB_FORM_BUTTON_CLEAR="Clear"
 JLIB_FORM_BUTTON_SELECT="Select"
+JLIB_FORM_CHANGE_FILE="Change file"
```
```suggestion
JLIB_FORM_CHANGE_FILE="Change File"
```

sergeytolkachyov committed:
Update language/en-GB/lib_joomla.ini

Co-authored-by: Quy Ton <quy@nomonkeybiz.com>

sergeytolkachyov committed:
Merge branch 'joomla:5.3-dev' into 5.3-dev

sergeytolkachyov committed:
`Change file` to `Change File`

sergeytolkachyov committed:
revert space changes in layout variables list

QuyTon commented on administrator/language/en-GB/plg_fields_media.ini:
```diff
@@ -9,4 +9,10 @@ PLG_FIELDS_MEDIA_PARAMS_DIRECTORY_LABEL="Directory"
 PLG_FIELDS_MEDIA_PARAMS_IMAGE_CLASS_LABEL="Image Class"
 PLG_FIELDS_MEDIA_PARAMS_PREVIEW_INLINE="Inline"
 PLG_FIELDS_MEDIA_PARAMS_PREVIEW_LABEL="Preview"
+PLG_FIELDS_MEDIA_PARAMS_TYPES_AUDIOS = "Audio"
```
Please remove spaces.

QuyTon commented on layouts/joomla/form/field/media.php:
```diff
@@ -85,9 +85,9 @@
     }
 
-    $width  = $previewWidth;
+    $width = $previewWidth;
```
Please revert. `=` should be aligned.

QuyTon commented on layouts/joomla/form/field/media.php:
```diff
@@ -148,13 +148,15 @@
-        'images'    => $imagesExt,
-        'audios'    => $audiosExt,
-        'videos'    => $videosExt,
+        'images' => $imagesExt,
```
Please revert. `=>` should be aligned.

QuyTon commented on plugins/fields/media/tmpl/media.php:
```diff
@@ -10,20 +10,37 @@
+use Joomla\CMS\HTML\HTMLHelper;
 use Joomla\CMS\Layout\LayoutHelper;
+use Joomla\CMS\Helper\MediaHelper;
+use Joomla\CMS\Language\Text;
```
```suggestion
use Joomla\CMS\Helper\MediaHelper;
use Joomla\CMS\HTML\HTMLHelper;
use Joomla\CMS\Language\Text;
use Joomla\CMS\Layout\LayoutHelper;
```
Alpha order.

sergeytolkachyov committed:
remove spaces

sergeytolkachyov committed:
code style

sergeytolkachyov committed:
Alpha order namespace import

brianteeman commented:
gets a bit confused if you select images + another type

![image](https://github.com/user-attachments/assets/14781095-32df-44f7-a6ed-5ea3e5316c83)

brianteeman commented:
php warnings

Create a field for documents and in an article select a pdf and save
The following php warning 
`[27-Feb-2025 14:24:45 UTC] PHP Warning:  mime_content_type(images/RL%20application%202017.pdf): Failed to open stream: No such file or directory in D:\repos\j51\libraries\src\Helper\MediaHelper.php on line 99`

Same for video
`Feb-2025 14:24:08 UTC] PHP Warning:  mime_content_type(images/sampledata/output.mp4): Failed to open stream: No such file or directory in D:\repos\j51\libraries\src\Helper\MediaHelper.php on line 99`

brianteeman commented:
From an accessibility perspective the default text for every link is "download"
This can result in multiple links on the same page with the exact same details - which is an accessibility failure

An option might be to do it the way we do a read more link. eg Download < Filename >

Tagging @chmst for their a11y input

brianteeman commented:
@viocassel how are you testing this?

sergeytolkachyov commented:
> php warnings
> 
> Create a field for documents and in an article select a pdf and save The following php warning `[27-Feb-2025 14:24:45 UTC] PHP Warning: mime_content_type(images/RL%20application%202017.pdf): Failed to open stream: No such file or directory in D:\repos\j51\libraries\src\Helper\MediaHelper.php on line 99`

I can't repeat the problem. Perhaps this is a MediaHelper and location issue on Windows.

![image](https://github.com/user-attachments/assets/d1d98e15-441b-48a5-b592-e99908d904e7)


> Same for video `Feb-2025 14:24:08 UTC] PHP Warning: mime_content_type(images/sampledata/output.mp4): Failed to open stream: No such file or directory in D:\repos\j51\libraries\src\Helper\MediaHelper.php on line 99`
> 
![image](https://github.com/user-attachments/assets/30ed4b19-4091-42ca-8dd1-d563563b98e7)

sergeytolkachyov commented:
> gets a bit confused if you select images + another type
> 
> ![image](https://private-user-images.githubusercontent.com/1296369/417631887-14781095-32df-44f7-a6ed-5ea3e5316c83.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA2NjY2MzMsIm5iZiI6MTc0MDY2NjMzMywicGF0aCI6Ii8xMjk2MzY5LzQxNzYzMTg4Ny0xNDc4MTA5NS0zMmRmLTQ0ZjctYTZlZC01ZWEzZTUzMTZjODMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDIyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAyMjdUMTQyNTMzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NWRjOTFhYWJkMmZhOTdlZTk0YzE3Y2RkZTY3MjZkY2ViYzk1M2RkMTgyNTRmYzEzODY0NWJlMDE2MDQ3YWRjYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.yniqee1b_HZimvPOQCPGdUzATph_0cEYmbIo3SOYWCE)

Yes, I agree. `Showon` comes to mind here, but there is no substring search so that you can show the rest of the fields based on the value of the main one.

sergeytolkachyov commented:
> for video and audio I would expect the output to be a player not a download.
> 
> see
> 
> https://developer.mozilla.org/en-US/docs/Web/HTML/Element/video https://developer.mozilla.org/en-US/docs/Web/HTML/Element/audio

To do this, I will need to add new layouts. However, even for pdf, you will want to show them as an `<object>` instead of a download link. Also, each file type may require its own additional data and settings specific to that particular file type. For videos, for example, the `poster` attribute or specifying a file in a different format (`mp4`, `webm` etc.). This can greatly complicate the logic of choosing the output layout. I assumed that at this point, the very fact that you can choose something other than pictures is already a good thing. And the rest can be solved using override.

sergeytolkachyov commented:
![image](https://github.com/user-attachments/assets/7cec0aab-8bd0-427c-920c-62e85f7a76b1)
What doesn't it like?

brianteeman commented:
If a user selects an image and it displays an image then when they select a video they would expect it to display a video etc overrides are for customising the output with a layout is fine but it should work out of the box without an override.

sergeytolkachyov commented:
> If a user selects an image and it displays an image then when they select a video they would expect it to display a video etc overrides are for customising the output with a layout is fine but it should work out of the box without an override.

So we need a `isVideo`, `isDocument` and `isAudio` methods for `MediaHelper`...  And check mime types too.

sergeytolkachyov committed:
Merge branch 'joomla:5.3-dev' into 5.3-dev

sergeytolkachyov committed:
Add audio and video layouts

sergeytolkachyov committed:
Link text field only for `documents` file type. `Audios` and `videos` removed due to its will render their own layouts

sergeytolkachyov committed:
Add support own layouts for `audios` and `videos` file types.

sergeytolkachyov commented:
> If a user selects an image and it displays an image then when they select a video they would expect it to display a video etc overrides are for customising the output with a layout is fine but it should work out of the box without an override.

I added this. Check it please.

sergeytolkachyov commented:
Testing instructions updated

sergeytolkachyov commented:
@brianteeman @hans2103 @QuyTon

sergeytolkachyov commented:
@gug2 @web-eau-net @hans2103 @viocassel can you test again?

brianteeman commented:
still get php warning on non-image media

example
```
Mar-2025 12:37:12 UTC] PHP Warning:  mime_content_type(images/rl.pdf): Failed to open stream: No such file or directory in D:\repos\j51\libraries\src\Helper\MediaHelper.php on line 99
[03-Mar-2025 12:37:12 UTC] PHP Warning:  mime_content_type(images/sampledata/output.mp4): Failed to open stream: No such file or directory in D:\repos\j51\libraries\src\Helper\MediaHelper.php on line 99
```

sergeytolkachyov commented:
> still get php warning on non-image media
> 
> example
> 
> ```
> Mar-2025 12:37:12 UTC] PHP Warning:  mime_content_type(images/rl.pdf): Failed to open stream: No such file or directory in D:\repos\j51\libraries\src\Helper\MediaHelper.php on line 99
> [03-Mar-2025 12:37:12 UTC] PHP Warning:  mime_content_type(images/sampledata/output.mp4): Failed to open stream: No such file or directory in D:\repos\j51\libraries\src\Helper\MediaHelper.php on line 99
> ```

Can you reproduce this issue on other Joomla installation?

QuyTon commented on layouts/joomla/form/field/media.php:
```diff
@@ -88,7 +88,7 @@
     $height = $previewHeight;
     $style  = ($width > 0) ? 'max-width:' . $width . 'px;' : '';
-    $style .= ($height > 0) ? 'max-height:' . $height . 'px;' : '';
+    $style  .= ($height > 0) ? 'max-height:' . $height . 'px;' : '';
```
Please revert.

QuyTon commented on layouts/joomla/form/field/media.php:
```diff
@@ -97,7 +97,7 @@
     $img = HTMLHelper::_('image', $src, Text::_('JLIB_FORM_MEDIA_PREVIEW_ALT'), $imgattr);
 
-    $previewImg      = '<div class="preview_img">' . $img . '</div>';
+    $previewImg = '<div class="preview_img">' . $img . '</div>';
```
Please revert.

sergeytolkachyov committed:
Merge branch 'joomla:5.3-dev' into 5.3-dev

sergeytolkachyov committed:
fix code style

sergeytolkachyov committed:
Merge branch '5.3-dev' into 5.3-dev

sergeytolkachyov commented:
@brianteeman

rdeutz commented:
All new feature have to go into 6.0, rebased the PR, sorry

sergeytolkachyov commented:
> All new feature have to go into 6.0, rebased the PR, sorry

https://joomlacommunity.cloud.mattermost.com/main/pl/ja16arwf7jrrxkqfmxs8dqo6dw

![image](https://github.com/user-attachments/assets/2db312cc-e504-46be-9c9c-b784b060a122)

Is there any generally accepted approach in this regard? In the seventh, yes, this PR should get into 6.0. However, I was informed in the chat that there was a chance to get into 5.3 (see screenshot). I decided to abandon a third-party plugin on my projects that implements this functionality and add it to the core. And I tried to make it just because I was informed that it was possible to add it to 5.3, which will be released soon (April). I'm not sorry for the work done, and I'm glad that after successful testing, I hope it will be included in the core.  But now you have to wait until the fall and use a third-party extension... And also make sure that maybe some other PR can do about the same thing or make changes that may affect the already tested and tested code.

sergeytolkachyov commented:
> not sure what you are asking me for. I tested this and reported a bug and there have been no changes by you since then so nothing will have changed

I can't reproduce this issue both on Windows local server and remote Linux server. I have tested 4 different Joomla installations. Here https://github.com/joomla/joomla-cms/pull/45013#issuecomment-2694365700 I show how the `MediaHelper` method works correctly for me. I asked 2 people to test it carefully and they also did not have such a problem. Therefore, I ask you to check the work on another Joomla installation. 

I also note that this problem occurs in the `MediaHelper` method, and it seems to me that a separate PR should be created to solve it so that everything is not mixed up in the current one.

brianteeman commented:
I will not be testing it on another server. I use one server for all my joomla tests and development. It's up to maintainers to decide if my test is good or not

sergeytolkachyov commented:
Can anyone else reproduce this bug?

QuyTon committed:
Merge branch '6.0-dev' into 5.3-dev

sergeytolkachyov committed:
Merge branch 'joomla:5.3-dev' into 5.3-dev

sergeytolkachyov committed:
Merge branch '6.0-dev' into 5.3-dev

brianteeman commented on administrator/language/en-GB/lib_joomla.ini:
```diff
@@ -277,6 +278,9 @@ JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_ALT_EMPTY_DESC="Decorative Image -
 JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_ALT_EMPTY_LABEL="No Description"
 JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_ALT_TEXT_LABEL="Image Description (Alt Text)"
 JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_IMAGEFILE_LABEL="Image"
+JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_LINKTEXT="Link text"
```
```suggestion
JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_LINKTEXT="Link Text"
```

brianteeman commented on language/en-GB/lib_joomla.ini:
```diff
@@ -276,6 +277,9 @@ JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_ALT_EMPTY_DESC="Decorative Image -
 JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_ALT_EMPTY_LABEL="No Description"
 JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_ALT_TEXT_LABEL="Image Description (Alt Text)"
 JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_IMAGEFILE_LABEL="Image"
+JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_LINKTEXT="Link text"
```
```suggestion
JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_LINKTEXT="Link Text"
```

brianteeman commented on layouts/joomla/html/audio.php:
```diff
@@ -0,0 +1,26 @@
+ * @package         Joomla.Site
+ * @subpackage      Layout
+ *
+ * @copyright   (C) 2021 Open Source Matters, Inc. <https://www.joomla.org>
```
```suggestion
 * @copyright       (C) 2025 Open Source Matters, Inc. <https://www.joomla.org>
```

brianteeman commented on layouts/joomla/html/video.php:
```diff
@@ -0,0 +1,40 @@
+ * @package         Joomla.Site
+ * @subpackage      Layout
+ *
+ * @copyright   (C) 2021 Open Source Matters, Inc. <https://www.joomla.org>
```
```suggestion
 * @copyright       (C) 2025 Open Source Matters, Inc. <https://www.joomla.org>
```

sergeytolkachyov committed:
Update administrator/language/en-GB/lib_joomla.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

sergeytolkachyov committed:
Update language/en-GB/lib_joomla.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

sergeytolkachyov committed:
Update layouts/joomla/html/audio.php

Co-authored-by: Brian Teeman <brian@teeman.net>

sergeytolkachyov committed:
Update layouts/joomla/html/video.php

Co-authored-by: Brian Teeman <brian@teeman.net>

sergeytolkachyov committed:
Merge branch '6.0-dev' into 5.3-dev

sergeytolkachyov committed:
Fix relative path to absolute

- Fix relative path to absolute for `MediaHelper::getMimeType()`
- add check `is_file`

laoneo commented:
This needs to be tested with a 3rd party extension like DPMedia or so.

laoneo commented on plugins/fields/media/tmpl/media.php:
```diff
@@ -10,20 +10,84 @@
-];
+$fileUrl = MediaHelper::getCleanMediaFieldValue($field->value['imagefile']);
+$fileFullPath = JPATH_SITE.DIRECTORY_SEPARATOR.$fileUrl;
+if(!\is_file($fileFullPath))
```
This will not work with external adapters which do read files from FTP filesystems, etc.

sergeytolkachyov commented:
> This needs to be tested with a 3rd party extension like DPMedia or so.

As I remember DP Media make copy of remote file to local filesystem. I'll try to use it

sergeytolkachyov commented:
> This needs to be tested with a 3rd party extension like DPMedia or so.

The `MediaHelper::getMimeTypes()` method does require an absolute path to the file and therefore there is a problem with remote files, but I cannot ignore this method because of the SVG files...

laoneo commented:
DPMedia makes only a copy to the local filesystem when the service doesn't hat public http based urls for images, like the FTP adapter. The Google implementation does link directly to the file on google drive. This is what the url attribute does then contain.

sergeytolkachyov commented:
> DPMedia makes only a copy to the local filesystem when the service doesn't hat public http based urls for images, like the FTP adapter. The Google implementation does link directly to the file on google drive. This is what the url attribute does then contain.

I try to detect external file url this way (`Uri::isInternal()` can't detect simple local path to file like `files/file.pdf`)

```php
$fileUrl = MediaHelper::getCleanMediaFieldValue($field->value['imagefile']);

// detect local file path
$isLocalFile = false;
if(empty((new Uri($fileUrl))->getHost()))
{
    $fileUrl = JPATH_SITE.DIRECTORY_SEPARATOR.$fileUrl;
    $isLocalFile = true;
}

if($isLocalFile && !\is_file($fileUrl))
{
    return;
}
// ....
```
Then check a Mime type
```php
if (MediaHelper::isImage($fileUrl) || ($isLocalFile && MediaHelper::getMimeType($fileUrl) === 'image/svg+xml')) {
    $options = [
        'src' => $field->value['imagefile'],
        'alt' => empty($field->value['alt_text']) && empty($field->value['alt_empty']) ? false : $field->value['alt_text'],
    ];

// ...
```
If we receive an external file, then we can determine whether it is an image, since the definition is simply based on the file extension. But we can't check its Mime type. Therefore, I have to set the logical `$isLocalFile` flag. This way we will be able to specify local SVG files in the field, but we will not be able to specify remote ones. 
Perhaps the best way would be to simply add the `svg` extension to the list of images in the `MediaHelper::isImage()` method, but I was guided by the code in the Joomla core, where the MIME type is additionally checked for SVG files.

sergeytolkachyov committed:
Remote file url support - Update media.php

Please note that not only local, but also remote ones can be specified in the field.

sergeytolkachyov committed:
Merge branch '6.0-dev' into 5.3-dev

QuyTon commented:
Please fix phpcs: https://ci.joomla.org/joomla/joomla-cms/84154/1/7

sergeytolkachyov committed:
Fix code style

sergeytolkachyov committed:
fix code style

sergeytolkachyov committed:
Merge branch '6.0-dev' into 5.3-dev

sergeytolkachyov committed:
Merge branch '6.0-dev' into 5.3-dev

sergeytolkachyov committed:
Merge branch '6.0-dev' into 5.3-dev

sergeytolkachyov committed:
Fix wrong file url

the path to the file was mistakenly passed, without taking into account the local or remote status.

sergeytolkachyov committed:
Merge branch '6.0-dev' into 5.3-dev

sergeytolkachyov committed:
Merge branch '6.0-dev' into 5.3-dev

sergeytolkachyov committed:
Merge branch '6.0-dev' into 5.3-dev

sergeytolkachyov committed:
Merge branch '6.0-dev' into 5.3-dev

gug2 commented:
I have tested this item :white_check_mark: successfully on 887919c83b8e9e517dd79febf63d7e3deaf25c5c<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/45013">issues.joomla.org/tracker/joomla-cms/45013</a>.</sub>

web-eau-net commented:
I have tested this item :white_check_mark: successfully on 887919c83b8e9e517dd79febf63d7e3deaf25c5c<br /><br />Successful test<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/45013">issues.joomla.org/tracker/joomla-cms/45013</a>.</sub>

sergeytolkachyov committed:
Merge branch '6.0-dev' into 5.3-dev

brianteeman commented:
> From an accessibility perspective the default text for every link is "download" This can result in multiple links on the same page with the exact same details - which is an accessibility failure
> 
> An option might be to do it the way we do a read more link. eg Download < Filename >
> 
> Tagging @chmst for their a11y input

This accessibility failure must be resolved

sergeytolkachyov committed:
Update lib_joomla.ini

If {filename} shortcode is found - it will be replaced

sergeytolkachyov committed:
Update media.php

added a `{filename}` short code processing

sergeytolkachyov committed:
Update media.php

added wasted basename()

sergeytolkachyov commented:
> > From an accessibility perspective the default text for every link is "download" This can result in multiple links on the same page with the exact same details - which is an accessibility failure
> > An option might be to do it the way we do a read more link. eg Download < Filename >
> > Tagging @chmst for their a11y input
> 
> This accessibility failure must be resolved

![image](https://github.com/user-attachments/assets/515f3e43-a9f9-4945-b788-657098157005)
Local pdf:
![image](https://github.com/user-attachments/assets/815512c5-c5fa-45f7-b2d9-39139317e439)
Remote pdf:
![image](https://github.com/user-attachments/assets/c9d1ab4f-e6ea-4f35-9cc0-6427d3d9b3f9)
Frontend
![image](https://github.com/user-attachments/assets/5df4fe92-fe12-4593-a7c4-c22d62a40007)

sergeytolkachyov committed:
Update media.php

fix code style

sergeytolkachyov committed:
Merge branch '6.0-dev' into 5.3-dev

sergeytolkachyov committed:
Merge branch '6.0-dev' into 5.3-dev

sergeytolkachyov commented:
@rdeutz can I do anything else for this PR?

sergeytolkachyov committed:
Merge branch '6.0-dev' into 5.3-dev

sergeytolkachyov committed:
Merge branch '6.0-dev' into 5.3-dev

sergeytolkachyov committed:
Merge branch '6.0-dev' into 5.3-dev

sergeytolkachyov committed:
Merge branch '6.0-dev' into 5.3-dev

sergeytolkachyov committed:
Update video.php

Fix typo in variable name

sergeytolkachyov committed:
Merge branch '6.0-dev' into 5.3-dev

sergeytolkachyov committed:
Merge branch '6.0-dev' into 5.3-dev

sergeytolkachyov commented:
@QuyTon can I do anything for this PR? I forgot make a separate branch for PR and then I can't do another one till this PR will be merged ((

sergeytolkachyov committed:
Merge branch '6.0-dev' into 5.3-dev

sergeytolkachyov committed:
Merge branch '6.0-dev' into 5.3-dev

sergeytolkachyov commented:
@QuyTon @rdeutz @brianteeman @laoneo Joomla 6.0.0 beta1 is coming in 2 days. **It means a feature freeze for J6.** This PR was succesfully tested and marked as `update requested` - so it should not be tested again. Is there anything else required for this PR? It has been opened in february...

brianteeman commented:
the people to tag are the release leads @softforge @Bodge-IT

sergeytolkachyov commented:
> the people to tag are the release leads @softforge @Bodge-IT

Thank you!

sergeytolkachyov committed:
Merge branch '6.0-dev' into 5.3-dev

Bodge-IT commented:
Hi @sergeytolkachyov, this only hit our radar on Sunday and we were locked by then. We had opinion this still needed resolutions.

sergeytolkachyov commented:
> Hi @sergeytolkachyov, this only hit our radar on Sunday and we were locked by then. We had opinion this still needed resolutions.

Can I do anything for this pr or it will wait for j6.1?

sergeytolkachyov commented:
> We are looking into it and will have a decision tomorrow, but appreciate the work you have done to date, thank you

If you have any questions, feel free to ask me. I myself see several potential ways to improve consistency, but this will generally require discussion with other developers before implementation.

HLeithner committed:
Merge branch '6.0-dev' into 5.3-dev

softforge commented:
Dear @sergeytolkachyov, first, thank you for all your work and patience so far. 
We had a look through the code and see there are still some comments to be resolved, and then testing as there have been commits since the last tests. 
Beta 1 has passed, so we cannot include it, but we would like to work with you over the next  few weeks to get any comments satisfied and then test it in PBF if it's ready for that or straight after, and to help you get it into 6.1 early so you are not waiting 6 months for a decision but it could be one of the first things ready for the 6.1 managers to merge

HLeithner commented:
This pull request has been automatically rebased to 6.1-dev.

sergeytolkachyov committed:
Merge branch '6.1-dev' into 5.3-dev

tecpromotion commented:
> 6.0.0 has been released. I'm waiting for feedback

We will be happy to take a look and give you feedback.

sergeytolkachyov committed:
Merge branch '6.1-dev' into 5.3-dev

sergeytolkachyov committed:
Merge branch '6.1-dev' into 5.3-dev

sergeytolkachyov committed:
Merge branch '6.1-dev' into 5.3-dev

sergeytolkachyov commented:
> > 6.0.0 has been released. I'm waiting for feedback
> 
> We will be happy to take a look and give you feedback.

Hello, just a reminder.

sergeytolkachyov committed:
Merge branch '6.1-dev' into 5.3-dev

sergeytolkachyov committed:
Merge branch '6.1-dev' into 5.3-dev

HLeithner commented on layouts/joomla/form/field/media.php:
```diff
@@ -155,6 +155,8 @@
     ]);
 }
 
+$imagesOnly = count($mediaTypeNames) == 1 && in_array('images', $mediaTypeNames);
```
```suggestion
$imagesOnly = count($mediaTypeNames) === 1 && in_array('images', $mediaTypeNames);
```

HLeithner commented on layouts/joomla/html/audio.php:
```diff
@@ -0,0 +1,26 @@
+
+use Joomla\Utilities\ArrayHelper;
+
+if (array_key_exists('src', $displayData) && !empty($displayData['src'])) {
```
```suggestion
if (!empty($displayData['src'])) {
```

sergeytolkachyov replied:
In some cases I got a `Undefined array key` with `!empty($displayData['src'])` warning that is why I did a double check

HLeithner replied:
that's not possible that's not how empty() works

example: https://3v4l.org/28m2H#v8.4.13

HLeithner commented on layouts/joomla/html/video.php:
```diff
@@ -0,0 +1,40 @@
+
+use Joomla\Utilities\ArrayHelper;
+
+if (array_key_exists('src', $displayData) && !empty($displayData['src'])) {
```
```suggestion
if (!empty($displayData['src'])) {
```

HLeithner commented on layouts/joomla/html/video.php:
```diff
@@ -0,0 +1,40 @@
+}
+
+$source = [];
+if (isset($displayData['source']) && is_array($displayData['source'])) {
```
```suggestion
if (!empty($displayData['source']) && is_array($displayData['source'])) {
```

HLeithner commented on libraries/src/Form/Field/AccessiblemediaField.php:
```diff
@@ -179,43 +191,55 @@ public function setup(\SimpleXMLElement $element, $value, $group = null)
-</form>
-XML;
-        $this->formsource = $xml;
+        $this->formsource = $xml->asXML();
```
What's the reason to change the xml string to function calls?

I see no benefit on this and the xml is easier to read than 20 calls to addAttribute.

sergeytolkachyov replied:
We can manipulate with result XML programatically. We add the fields `alt_text` and `alt_empty` only if `images` file type has been selected. The same is for `documents` type - we will display the additional field with `Download {filename}` only if `documents` type has been selected. But if we have a combination with `images` + `videos` or `videos` + `audios` only - there wouldn't be shown. 

I think using `SimpleXMLElement` is much better then manipulation with string

HLeithner replied:
it's also much slower, anyway you do some language translation already here in your code which is the wrong position. this is always done when we output the strings.

and you convert it back to a string? I don't like this please change it back to a string. I think you can do this easily also by concat a string.

sergeytolkachyov replied:
> What's the reason to change the xml string to function calls?
> 
> I see no benefit on this and the xml is easier to read than 20 calls to addAttribute.

I remembered the reason why I used `SimpleXMLElement`. In the admin panel, the `default` value from the language constant `JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_LINKTEXT_DEFAULT_VALUE` is inserted in the field. And if the `Text` class works for hint attributes, then it doesn't for the `default` attribute. While we need a localized default value in it.

This led to the error that @brianteeman pointed out below in his video.

When there are long breaks in working on the code, the context is forgotten...

HLeithner reviewed (changes requested):
We would like to have this in 6.1, it's not perfect that you see all fields for all types but ok for now. If you like you might write some javascript to remove fields which doesn't fit to the selected file type.

I have some code comments please check all of them.

thanks

sergeytolkachyov committed:
Update layouts/joomla/form/field/media.php

Co-authored-by: Harald Leithner <leithner@itronic.at>

HLeithner commented:
> > it's not perfect that you see all fields for all types but ok for now
> 
> if you selected a `documents` and `images` types together then for `images` you need this fields, but not need for `documents`.
> 
> So I think it would be better to rename a `imagefile` field input to `file` but it is a b/c break.

wouldn't change it to `imagefile` now could be done in a `x.0` release but not in a minor.

sergeytolkachyov committed:
Revert SimpleXMLElement to string

HLeithner commented:
@sergeytolkachyov thanks for the revert, could you fix the code style? 
Since we use heredoc you can indent the full block so it doesn't start left instead under the variable name. But I can life with the existing indent. Just fix the cs please

sergeytolkachyov commented:
> @sergeytolkachyov thanks for the revert, could you fix the code style? Since we use heredoc you can indent the full block so it doesn't start left instead under the variable name. But I can life with the existing indent. Just fix the cs please

That is why I didn't like heredoc syntax)) One more space and all is crashed))) 

I'll fix it

sergeytolkachyov committed:
Update AccessiblemediaField.php fix code style

sergeytolkachyov committed:
Update AccessiblemediaField.php remove unused namespace

tecpromotion commented on libraries/src/Form/Field/AccessiblemediaField.php:
```diff
@@ -212,9 +228,26 @@ public function setup(\SimpleXMLElement $element, $value, $group = null)
+		<field
+				name="linktext"
+				type="text"
+				label="JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_LINKTEXT"
```
```suggestion
				label="JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_LINKTEXT_LABEL"
```

tecpromotion committed:
Add label for Link Text parameter in lib_joomla.ini

tecpromotion committed:
Add label for Link Text parameter in lib_joomla.ini

tecpromotion committed:
Apply suggestions from code review

tecpromotion committed:
Merge branch '6.1-dev' into 5.3-dev

brianteeman commented on language/en-GB/lib_joomla.ini:
```diff
@@ -276,6 +277,9 @@ JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_ALT_EMPTY_DESC="Decorative Image -
 JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_ALT_EMPTY_LABEL="No Description"
 JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_ALT_TEXT_LABEL="Image Description (Alt Text)"
 JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_IMAGEFILE_LABEL="Image"
+JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_LINKTEXT_DEFAULT_VALUE="Download"
```
why is this different to the admin string?

tecpromotion replied:
Thanks @brianteeman
There is no reason. Feel free to make a suggestion.

HLeithner commented:
@sergeytolkachyov can you address the things brian mentioned? about the "image" label for all types and img-tag even for not images? and the placeholder text? I don't care about the additional fields in the "all" types variant, where you might add an image description for a audio file, we can't solve this easily so I would accept this for now. But the other things should be addressed before merging.

thanks

brianteeman commented on administrator/language/en-GB/plg_fields_media.ini:
```diff
@@ -9,4 +9,10 @@ PLG_FIELDS_MEDIA_PARAMS_DIRECTORY_LABEL="Directory"
 PLG_FIELDS_MEDIA_PARAMS_PREVIEW_INLINE="Inline"
 PLG_FIELDS_MEDIA_PARAMS_PREVIEW_LABEL="Preview"
+PLG_FIELDS_MEDIA_PARAMS_TYPES_AUDIOS="Audio"
+PLG_FIELDS_MEDIA_PARAMS_TYPES_DESCRIPTION="You can configure a file types in Media Manager settings"
```
```suggestion
PLG_FIELDS_MEDIA_PARAMS_TYPES_DESCRIPTION="You can configure the allowed file types in the Media Manager settings."
```

correcting the english and following the styleguide

sergeytolkachyov committed:
Update administrator/language/en-GB/plg_fields_media.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

sergeytolkachyov committed:
Update lib_joomla.ini

Make language strings the same with the admin file

sergeytolkachyov committed:
Add AccessiblemediaField getInput method

Add `getInput` method to set `linktext` rendered `default` value and improve `label` handling for `media` field

sergeytolkachyov committed:
Update media labels in lib_joomla.ini

Add media types language strings - images, audios, documents, videos. Includ types to media field label

sergeytolkachyov committed:
Update media labels in lib_joomla.ini

add language strings

sergeytolkachyov committed:
Change lang strings order

sergeytolkachyov committed:
Fix formatting in AccessiblemediaField.php

sergeytolkachyov committed:
Fix code style in AccessiblemediaField.php

sergeytolkachyov committed:
Merge branch '6.1-dev' into 5.3-dev

sergeytolkachyov committed:
Fix formatting of labels array initialization

sergeytolkachyov committed:
Add use statement for Joomla CMS Language Text

sergeytolkachyov committed:
Split one multitypes to 4 different

sergeytolkachyov committed:
Fix code style

sergeytolkachyov committed:
Fix code style

sergeytolkachyov committed:
Merge branch '6.1-dev' into 5.3-dev

brianteeman commented on plugins/fields/media/tmpl/audio.php:
```diff
@@ -0,0 +1,46 @@
+ * @package     Joomla.Plugin
+ * @subpackage  Fields.Media
+ *
+ * @copyright   (C) 2016 Open Source Matters, Inc. <https://www.joomla.org>
```
```suggestion
 * @copyright   (C) 2026 Open Source Matters, Inc. <https://www.joomla.org>
```

brianteeman commented on administrator/language/en-GB/lib_joomla.ini:
```diff
@@ -272,11 +273,20 @@ JLIB_FORM_FIELD_INVALID_MIN_TIME="The time you entered is before the minimum tim
+JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_MEDIA_TYPE_DOCUMENTS="Documents"
+JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_MEDIA_TYPE_IMAGES="Images"
+JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_MEDIA_TYPE_VIDEOS="Videos"
+JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_VIDEO_POSTER_LABEL="Video poster"
```
```suggestion
JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_VIDEO_POSTER_LABEL="Video Poster"
```

brianteeman commented on layouts/joomla/html/audio.php:
```diff
@@ -0,0 +1,26 @@
+ * @package         Joomla.Site
+ * @subpackage      Layout
+ *
+ * @copyright       (C) 2025 Open Source Matters, Inc. <https://www.joomla.org>
```
```suggestion
 * @copyright       (C) 2026 Open Source Matters, Inc. <https://www.joomla.org>
```

brianteeman replied:
```suggestion
 * @copyright       (C) 2026 Open Source Matters, Inc. <https://www.joomla.org>
```

sergeytolkachyov committed:
Fix code style

brianteeman commented on layouts/joomla/html/video.php:
```diff
@@ -0,0 +1,40 @@
+ * @package         Joomla.Site
+ * @subpackage      Layout
+ *
+ * @copyright       (C) 2025 Open Source Matters, Inc. <https://www.joomla.org>
```
```suggestion
 * @copyright       (C) 2026 Open Source Matters, Inc. <https://www.joomla.org>
```

brianteeman replied:
```suggestion
 * @copyright       (C) 2026 Open Source Matters, Inc. <https://www.joomla.org>
```

sergeytolkachyov committed:
Merge remote-tracking branch 'origin/5.3-dev' into 5.3-dev

brianteeman commented on libraries/src/Form/Field/AccessiblemediaField.php:
```diff
@@ -212,13 +231,63 @@ public function setup(\SimpleXMLElement $element, $value, $group = null)
+     *
+     * @since   6.1.0
+     */
+    protected function getInput()
```
```suggestion
     * @since   __DEPLOY_VERSION__
```

brianteeman commented on plugins/fields/media/tmpl/document.php:
```diff
@@ -0,0 +1,43 @@
+ * @package     Joomla.Plugin
+ * @subpackage  Fields.Media
+ *
+ * @copyright   (C) 2016 Open Source Matters, Inc. <https://www.joomla.org>
```
```suggestion
 * @copyright   (C) 2026 Open Source Matters, Inc. <https://www.joomla.org>
```

brianteeman commented on plugins/fields/media/tmpl/video.php:
```diff
@@ -0,0 +1,55 @@
+ * @package     Joomla.Plugin
+ * @subpackage  Fields.Media
+ *
+ * @copyright   (C) 2016 Open Source Matters, Inc. <https://www.joomla.org>
```
```suggestion
 * @copyright   (C) 2026 Open Source Matters, Inc. <https://www.joomla.org>
```

sergeytolkachyov committed:
Fix PHPDoc @since tag and code style

sergeytolkachyov committed:
Add lost poster field

sergeytolkachyov committed:
The only number in the field label is for a single file type, while maintaining the file type enumeration if there are several specified.

sergeytolkachyov committed:
fix code style

sergeytolkachyov committed:
fix code style

sergeytolkachyov committed:
fix code style

brianteeman commented on libraries/src/Form/Field/AccessiblemediaField.php:
```diff
@@ -212,13 +231,88 @@ public function setup(\SimpleXMLElement $element, $value, $group = null)
+				label="JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_LINKTEXT_LABEL"
+				description="JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_LINKTEXT_DESC"
+				filter="string"
+				hint="JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_LINKTEXT_DEFAULT_VALUE"
```
```suggestion
			name="linktext"
			type="text"
			label="JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_LINKTEXT_LABEL"
			description="JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_LINKTEXT_DESC"
			filter="string"
			hint="JLIB_FORM_FIELD_PARAM_ACCESSIBLEMEDIA_PARAMS_LINKTEXT_DEFAULT_VALUE"
```

sergeytolkachyov committed:
Update layouts/joomla/html/audio.php

Co-authored-by: Brian Teeman <brian@teeman.net>

sergeytolkachyov committed:
Update libraries/src/Form/Field/AccessiblemediaField.php

Co-authored-by: Brian Teeman <brian@teeman.net>

sergeytolkachyov committed:
Update layouts/joomla/html/video.php

Co-authored-by: Brian Teeman <brian@teeman.net>

sergeytolkachyov commented:
> Example of what I mean in the xml where you need to remove the spaces and replace with a tab
> 
> <img alt="image" width="637" height="275" src="https://private-user-images.githubusercontent.com/1296369/547836644-82e8a61c-2a58-4aac-8696-8ed593baa448.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzA3NDQzMTcsIm5iZiI6MTc3MDc0NDAxNywicGF0aCI6Ii8xMjk2MzY5LzU0NzgzNjY0NC04MmU4YTYxYy0yYTU4LTRhYWMtODY5Ni04ZWQ1OTNiYWE0NDgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDIxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjAyMTBUMTcyMDE3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZGZjZGQ2NWU3MjE0YjkyNjU5ODhlYjEwZDk0MmE4ZjczMTEyZGY4MGZjMzI5NTdkNDZjNDQwMzg5MzdiZDJlYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Mq-B1imQtgA34SlP6eFSjJISmihU_3o1KhYlrjYs5qs">

Thank you. Ok, tomorrow.

sergeytolkachyov committed:
fix tabs
